cpp/ql/lib/semmle/code/cpp/internal/Overlay.qll

```diff
@@ -0,0 +1,64 @@
+/**
+ * Defines entity discard predicates for C++ overlay analysis.
+ */
+
+/**
+ * Holds always for the overlay variant and never for the base variant.
+ * This local predicate is used to define local predicates that behave
+ * differently for the base and overlay variant.
+ */
+overlay[local]
+predicate isOverlay() { databaseMetadata("isOverlay", "true") }
+
+/** Gets the file path for a location. */
+overlay[local]
+private string getLocationFilePath(@location_default loc) {
+  exists(@file file | locations_default(loc, file, _, _, _, _) | files(file, result))
+}
+
+overlay[local]
+private class DiscardableEntityBase extends @element {
+  /** Gets the path to the file in which this element occurs. */
+  abstract string getFilePath();
+
+  /** Holds if this element exists in the base variant. */
+  predicate existsInBase() { not isOverlay() }
+
+  /** Gets a textual representation of this discardable element. */
+  string toString() { none() }
+}
+
+/**
+ * Discard an entity from the base if all its locations are in changed files.
+ * Entities with at least one location in an unchanged file are kept.
+ */
+overlay[discard_entity]
+private predicate discardEntity(@element e) {
+  e =
+    any(DiscardableEntityBase de |
+      de.existsInBase() and
+      overlayChangedFiles(de.getFilePath()) and
+      // Only discard if ALL file paths are in changed files
+      forall(string path | path = de.getFilePath() | overlayChangedFiles(path))
```

Comment:
It seems that there are two cases here:

* Elements that have a location in the relation that define them (like variable declarations). In this case I don't think we need the `forall`, and we can do what you were doing before.
* Elements that do not come with their own location, but where the location can come from one or more other elements (this is probably just variables, functions, and types?).

Do we want to make that distinction more explicit?